### PR TITLE
feat/file-search-by-name

### LINKED ISSUE #2368: Allow to search across all Namespace Files (rather than only across currently-open files)
1. Search by file name
2. Search by flow source

Search by file's content is out of scope for now for several reasons (performance, privacy)

### PR DESCRIPTION
**Pull request by brian-mulier-p**
closes #2368

### CODE REVIEW

File: core/src/main/java/io/kestra/core/storages/StorageInterface.java [hunk 1]
```diff
@@ -30,6 +30,13 @@ public interface StorageInterface {
     @Retryable(includes = {IOException.class}, excludes = {FileNotFoundException.class})
     InputStream get(String tenantId, URI uri) throws IOException;
 
+    /**
+     * Returns all existing file paths' that start with the given prefix.
+     * @return relative paths starting from prefix
```
Comment (loicmathieu) [inline core/src/main/java/io/kestra/core/storages/StorageInterface.java:35]: The second line is misleading, does it retruns path or path relative to the preifx?
I think it should return full URI

Comment (loicmathieu) [reply core/src/main/java/io/kestra/core/storages/StorageInterface.java:35]: So the test shows that it returns relative path which is strange.
Looking at the interface of the LocalStorage I think it should returns a List<URI> instead, containing the full internal storage URI.
This if you need a list of string without the prefix it is something related to NamespaceFiles so it must be adapted inside the NamespaceFile service.
Otherwise, this method will not be reusable for something else.

File: storage-local/src/main/java/io/kestra/storage/local/LocalStorage.java [hunk 2]
```diff
@@ -50,6 +50,17 @@ public InputStream get(String tenantId, URI uri) throws IOException {
         );
     }
 
+    @Override
+    public List<String> filePathsByPrefix(String tenantId, URI prefix) throws IOException {
```
Comment (loicmathieu) [inline storage-local/src/main/java/io/kestra/storage/local/LocalStorage.java:54]: I don't understand in what this is different from the existing method that list all files for a namespace?

**Review by loicmathieu — CHANGES_REQUESTED** [review]

File: storage-local/src/main/java/io/kestra/storage/local/LocalStorage.java (hunk 2 above)
Comment (brian-mulier-p) [reply storage-local/src/main/java/io/kestra/storage/local/LocalStorage.java:54]: It's recursive. Its goal is to be able to get every file (note the word file, we don't want directories in here) with their path.
Following what you said above, I can definitely return the entire URI if that's preferable and do the change in NamespaceFile service

File: core/src/main/java/io/kestra/core/storages/StorageInterface.java (hunk 1 above)
Comment (brian-mulier-p) [reply core/src/main/java/io/kestra/core/storages/StorageInterface.java:35]: replied in the other comment

File: storage-local/src/main/java/io/kestra/storage/local/LocalStorage.java (hunk 2 above)
Comment (loicmathieu) [reply storage-local/src/main/java/io/kestra/storage/local/LocalStorage.java:54]: > Following what you said above, I can definitely return the entire URI if that's preferable and do the change in NamespaceFile service

Yes, full URIs please, internal storage always use full URIs

**Comment by brian-mulier-p** [comment]
I made the full URI change

File: core/src/test/java/io/kestra/core/storage/StorageTestSuite.java [hunk 3]
```diff
@@ -112,6 +112,38 @@ void getFileNotFound() {
...
+        res = storageInterface.filesByPrefix(null, URI.create("/namespace/folder"));
+        assertThat(res, containsInAnyOrder(
+            URI.create("kestra:///namespace/folder/file.txt"),
+            URI.create("kestra:///namespace/folder/some.yaml"),
+            URI.create("kestra:///namespace/folder/sub/script.py")
+        ));
+
+        res = storageInterface.filesByPrefix(null, URI.create("/namespace/folder/sub"));
+        assertThat(res, containsInAnyOrder(URI.create("kestra:///namespace/folder/sub/script.py")));
+    }
```
Comment (loicmathieu) [inline core/src/test/java/io/kestra/core/storage/StorageTestSuite.java:145]: Please add a test for when there is no response (should return an empty list)

File: webserver/src/main/java/io/kestra/webserver/controllers/NamespaceFileController.java [hunk 4]
```diff
@@ -54,6 +54,21 @@ public class NamespaceFileController {
...
 
+    @ExecuteOn(TaskExecutors.IO)
+    @Get(uri = "{namespace}/files/search", produces = MediaType.TEXT_JSON)
+    @Operation(tags = {"Files"}, summary = "Find files which path contain the given string in their URI")
+    public List<String> search(
+        @Parameter(description = "The namespace id") @PathVariable String namespace,
+        @Parameter(description = "The string the file path should contain") @QueryValue String q
+    ) throws IOException, URISyntaxException {
+        URI baseNamespaceFilesUri = toNamespacedStorageUri(namespace, null);
+        return Stream.concat(
```
Comment (loicmathieu) [inline webserver/src/main/java/io/kestra/webserver/controllers/NamespaceFileController.java:65]: Why do you concat with staticFiles and what is this statifFiles?

**Review by loicmathieu — CHANGES_REQUESTED** [review]
I only saw a new API endpoint that is not yet used.
Isn't it any way to test it from the UI?

File: webserver/src/main/java/io/kestra/webserver/controllers/NamespaceFileController.java (hunk 4 above)
Comment (brian-mulier-p) [reply webserver/src/main/java/io/kestra/webserver/controllers/NamespaceFileController.java:65]: For easier maintainability and flexibility I added a static file concept for VSCode editor. Those files are served from static resources of the webserver and the goal is to be able as a software to provide some read-only files to the user. This can be used to guide him (for eg. we currently have a getting-started file which appear on VSCode editor but isn't editable, it is served statically and acts like a README for editor usage)

Comment (loicmathieu) [reply webserver/src/main/java/io/kestra/webserver/controllers/NamespaceFileController.java:65]: Well, I'm not sure I see the point of including them in the search but anyway, OK

**Review by loicmathieu — APPROVED** [review]
LGTM.

QA via the API as the currently merged extension is not capable of calling it.
